Modernize CMake CUDA to 3.25+ (native arch, device LTO) (#4773)
AMReX-Codes/amrex: Docs/sphinx_documentation/source/GPU.rst

   \end{center}

The target architecture(s) to build for are selected via the standard CMake option
``-DCMAKE_CUDA_ARCHITECTURES=<arch>`` (or the ``CUDAARCHS`` environment variable), where
``<arch>`` is one or more `CUDA compute capabilities <https://developer.nvidia.com/cuda-gpus>`_
given as integers without the dot, e.g., ``80`` for compute capability 8.0. Multiple
architectures can be set as a semicolon-separated list, e.g.,

> ``-DCMAKE_CUDA_ARCHITECTURES="80;90"``; building for multiple architectures generally

Author: ```suggestion
``-DCMAKE_CUDA_ARCHITECTURES="80;90"``. Building for multiple architectures generally
```

results in a larger library and longer build times. The special values ``native`` (the
default), ``all`` and ``all-major``, as well as the ``-real``/``-virtual`` and ``<NN>a``
suffixes documented for CMake's
`CUDA_ARCHITECTURES <https://cmake.org/cmake/help/latest/prop_tgt/CUDA_ARCHITECTURES.html>`__
target property, are supported.

.. highlight:: console

::

   cmake [options] -DAMReX_GPU_BACKEND=CUDA -DCMAKE_CUDA_ARCHITECTURES=80 /path/to/amrex/source

If no architecture is specified, AMReX defaults to ``native``, which builds for the GPU(s)
installed in the machine running CMake. This requires a GPU to be visible at configuration

> time; on GPU-less machines (e.g., HPC login nodes or CI runners) an explicit architecture

Author: ```suggestion
time. On machines without GPUs (e.g., HPC login nodes or CI runners) an explicit architecture
```

must be provided.

.. note::

   The legacy ``-DAMReX_CUDA_ARCH`` option and the ``AMREX_CUDA_ARCH`` environment variable
   are deprecated but still honored (with a warning): they map to ``CMAKE_CUDA_ARCHITECTURES``
   and the standard ``CUDAARCHS`` environment variable, respectively, and keep their historical
   precedence. Please migrate to the standard CMake variables.

Setting ``-DAMReX_CUDA_LTO=ON`` enables CUDA device link-time optimization. This requires

> relocatable device code (``AMReX_GPU_RDC=ON``, the default) and CMake 3.25 or newer.

Author: We require CMake 3.25+ now:
```suggestion
relocatable device code (``AMReX_GPU_RDC=ON``, the default).
```

   \end{center}

The target architecture(s) to build for are selected via the standard CMake option

> ``-DCMAKE_CUDA_ARCHITECTURES=<arch>`` (or the ``CUDAARCHS`` environment variable), where

Author: Add an explicit section about environment hints to set for CMake (all standard, but good to say here):

```bash
# For Cray/HPE machines: necessary to use CUDA-Aware MPI and run a job
export CRAY_ACCEL_TARGET=nvidia80

# example: overwrite the auto-detection to optimize CUDA compilation for A100
export CUDAARCHS=80

# example: optimize CPU microarchitecture for AMD EPYC 3rd Gen (Milan/Zen3)
export CXXFLAGS="-march=znver3"
export CFLAGS="-march=znver3"

# CMake compiler environment hints
export CC=$(which gcc)       # or $(which cc) for Cray/HPE
export CXX=$(which g++)      # or $(which CC) for Cray/HPE
export FC=$(which gfortran)  # or $(which ftn) for Cray/HPE
export CUDACXX=$(which nvcc)
export CUDAHOSTCXX=${CXX}
```

Author: added

Author: added

installed in the machine running CMake. This requires a GPU to be visible at configuration
time. On machines without GPUs (e.g., HPC login nodes or CI runners) an explicit architecture
must be provided.

.. note::

   The legacy ``-DAMReX_CUDA_ARCH`` option and the ``AMREX_CUDA_ARCH`` environment variable
   are deprecated but still honored (with a warning): they map to ``CMAKE_CUDA_ARCHITECTURES``
   and the standard ``CUDAARCHS`` environment variable, respectively, and keep their historical
   precedence. Please migrate to the standard CMake variables.

Setting ``-DAMReX_CUDA_LTO=ON`` enables CUDA device link-time optimization. This requires
relocatable device code (``AMReX_GPU_RDC=ON``, the default). Device LTO is currently
available only with the NVIDIA ``nvcc`` compiler; with a Clang ``-x cuda`` compiler it is
gracefully disabled with a warning.

**Note that AMReX supports NVIDIA GPU architectures with compute capability 6.0 or higher and
CUDA Toolkit version 12.2 or higher.**

**Environment hints.** All of AMReX's CMake configuration can also be driven through standard
CMake and compiler environment variables. These are especially handy on HPC systems and in CI,

> where they are typically exported in a job or module script before invoking ``cmake``:

Author: ```suggestion
where they are typically exported in a job or module script before invoking ``cmake``.
Example on a Cray/HPE machine:
```

.. code-block:: bash

>    # For Cray/HPE machines: necessary to use CUDA-Aware MPI and run a job

Author: ```suggestion
   # For Cray/HPE machines: necessary to use CUDA-Aware MPI and run a job
   # Note: also sometimes already set by a loaded module.
```

   export CRAY_ACCEL_TARGET=nvidia80

   # example: overwrite the auto-detection to optimize CUDA compilation for A100
   export CUDAARCHS=80

>    # example: optimize CPU microarchitecture for AMD EPYC 3rd Gen (Milan/Zen3)

Author: ```suggestion
   # overwrite the auto-detection to optimize CUDA compilation for A100
   export CUDAARCHS=80

   # optimize CPU (host) microarchitecture for AMD EPYC 3rd Gen (Milan/Zen3)
```

   export CXXFLAGS="-march=znver3"
   export CFLAGS="-march=znver3"

>    # CMake compiler environment hints

Author: ```suggestion
   # CMake compiler selection
```
